packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -492,37 +494,63 @@ class OAuthClient extends CozyStackClient {
    * @returns {Promise} A promise that resolves with a new AccessToken object
    */
   async refreshToken() {
-    if (!this.isRegistered()) throw new NotRegisteredException()
-    if (!this.token) throw new Error('No token to refresh')
+    if (!this.isRegistered()) {
+      throw new NotRegisteredException()
+    }
+    if (!this.token) {
+      throw new Error('No token to refresh')
+    }
 
     const data = {
       grant_type: 'refresh_token',
       refresh_token: this.token.refreshToken,
       client_id: this.oauthOptions.clientID,
       client_secret: this.oauthOptions.clientSecret
     }
+    try {
+      const result = await this.fetchJSONWithCurrentToken(
+        'POST',
+        '/auth/access_token',
+        this.dataToQueryString(data),
+        {
+          headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+        }
+      )
 
-    const result = await super.fetchJSON(
-      'POST',
-      '/auth/access_token',
-      this.dataToQueryString(data),
-      {
-        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+      const newToken = new AccessToken({
+        refresh_token: this.token.refreshToken,
+        ...result
+      })
+
+      this.setToken(newToken)
+
+      if (this.onTokenRefresh && typeof this.onTokenRefresh === 'function') {
+        this.onTokenRefresh(newToken)
       }
-    )
 
-    const newToken = new AccessToken({
-      refresh_token: this.token.refreshToken,
-      ...result
-    })
+      return newToken
+    } catch (e) {
+      if (errors.UNREGISTERED_CLIENT.test(e.message)) {
+        this.onRevocationChange(true)
 
-    this.setToken(newToken)
+        throw new Error(
+          'Client has been revoked or does not exist. Please authenticate again.'
+        )
+      }
 
-    if (this.onTokenRefresh && typeof this.onTokenRefresh === 'function') {
-      this.onTokenRefresh(newToken)
-    }
+      if (
+        errors.EXPIRED_TOKEN.test(e.message) ||
+        errors.INVALID_TOKEN.test(e.message) ||
+        errors.INVALID_TOKEN_ALT.test(e.message)
+      ) {
+        const revoked = await this.checkForRevocation()
 
-    return newToken
+        if (revoked) {
```

Comment:
same here, I think that checkForRevocation() should throw the real error.

Reply:
ok, then the initial fetcherror? as I mentioned earlier, I think we should add some more context info maybe. Like adding we think the client is revoked (I mean it's 100% sure it doesn't exist, but it's either revoked or non-existing)

Reply:
- [ ] Updated